snapcraft/plugins/_python/_pip.py

```diff
@@ -0,0 +1,339 @@
+        :param str stage_dir: Path to the staging area
+
+        :raises MissingPythonCommandError: If no python could be found in the
+                                           staging or part's install area.
+        """
+        self._python_major_version = python_major_version
+        self._part_dir = part_dir
+        self._install_dir = install_dir
+        self._stage_dir = stage_dir
+
+        self._python_package_dir = os.path.join(
+            self._part_dir, 'python-packages')
+        os.makedirs(self._python_package_dir, exist_ok=True)
+
+        self._python_command = get_python_command(
+            self._python_major_version, stage_dir=self._stage_dir,
+            install_dir=self._install_dir)
+        self._python_home = get_python_home(
+            self._python_major_version, stage_dir=self._stage_dir,
+            install_dir=self._install_dir)
+
+        self._setup()
+
+    def _setup(self):
+        # Check to see if we have our own pip, yet. If not, we need to use the
+        # pip on the host (installed via build-packages) to grab our own.
+        if not self._is_pip_installed():
+            logger.info('Fetching pip, setuptools, and wheel...')
+            real_python_home = self._python_home
+
+            # Make sure we're using pip from the host
+            self._python_home = os.path.join(os.path.sep, 'usr')
+
+            # Using the host's pip, install our own pip and other tools we
+            # need.
+            self.download({'pip', 'setuptools', 'wheel'})
+            self.install({'pip', 'setuptools', 'wheel'}, ignore_installed=True)
+
+            # Now that we have our own pip, reset the python home
+            self._python_home = real_python_home
+
+    def _is_pip_installed(self):
+        try:
+            self._run([], stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as e:
+            output = e.output.decode(sys.getfilesystemencoding()).strip()
+            if 'no module named pip' in output.lower():
+                return False
+            else:
+                raise e
+
+        return True
+
+    def download(self, packages, *, setup_py_dir=None, constraints=None,
+                 requirements=None, process_dependency_links=False):
+        """Download packages into cache, but do not install them.
+
+        :param iterable packages: Packages to download from index.
+        :param str setup_py_dir: Directory containing setup.py.
+        :param iterable constraints: Collection of paths to constraints files.
+        :param iterable requirements: Collection of paths to requirements
+                                      files.
+        :param boolean process_dependency_links: Enable the processing of
+                                                 dependency links.
+        """
+        args = []
+        cwd = None
+        if setup_py_dir:
+            args.append('.')
+            cwd = setup_py_dir
+
+        _process_common_args(
```

Comment:
this would convey more readability if in the end we assign `args` to what this returns.

Reply:
Yeah, stopped passing args at all, and assigned.